example_test.go

```diff
@@ -0,0 +1,102 @@
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package fx_test
+
+import (
+	"log"
+	"net/http"
+	"os"
+	"time"
+
+	"go.uber.org/fx"
+)
+
+func NewLogger() *log.Logger {
+	logger := log.New(os.Stdout, "[Example] " /* prefix */, 0 /* flags */)
+	logger.Print("Executing NewLogger.")
+	return logger
+}
+
+func NewHandler(logger *log.Logger) http.Handler {
+	logger.Print("Executing NewHandler.")
+	return http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
+		logger.Print("Got a request.")
+		w.Write([]byte("Your lucky number is 42."))
+	})
+}
+
+func NewMux(lc fx.Lifecycle, logger *log.Logger) *http.ServeMux {
+	logger.Print("Executing NewMux.")
+	mux := http.NewServeMux()
+	server := &http.Server{
+		Addr:    ":0",
+		Handler: mux,
+	}
+	// If NewMux is called, we know that someone is using the mux. In that case,
+	// start up and shut down an HTTP server with the application.
+	lc.Append(fx.Hook{
+		OnStart: func() error {
+			logger.Print("Starting HTTP server.")
+			go server.ListenAndServe()
+			return nil
+		},
+		OnStop: func() error {
+			logger.Print("Stopping HTTP server.")
+			return server.Close()
+		},
+	})
+
+	return mux
+}
+
+func Register(mux *http.ServeMux, h http.Handler) {
+	mux.Handle("/", h)
+}
+
+func Example() {
+	app := fx.New(
+		// Provide all the constructors we need.
+		fx.Provide(NewLogger, NewHandler, NewMux),
+		// Before starting, register the handler. This forces resolution of all
+		// the types in the container. Since the mux is now being used, its
+		// startup hook gets registered and the application includes an HTTP
+		// server.
+		fx.Invoke(Register),
+	)
+
+	// In a real application, we could just use app.Run() here. Since we don't
+	// want this example to run forever, we'll use Start and Stop.
+	if err := app.Start(fx.Timeout(time.Second)); err != nil {
+		log.Fatal(err)
+	}
+
+	// Normally, we'd block here with <-app.Done().
```

Comment:
Well if the cost of doing so outweighs the benefits, then there is no reason to bother. Was just an idea, I don't even know if it's going to show up well in godoc